### PR TITLE
Unifying manifest target generation

### PR DESCRIPTION
**Pull request by kwridan**
Part of: #205 

### Short description 📝

When generating projects, Tuist generates user defined targets as well as a manifest target to host the project manifest (`Project.swift` file). Currently the generator has a set of dedicated methods to generating the manifest target and its config. This could be simplified by re-using the existing target generation code for the other targets.

### Solution 📦

When creating the project models, we could add one more target (the manifest target) with the necessary custom settings. 

### Implementation 👩‍💻👨‍💻

- Introduce a manifest target generator that generates the necessary `Target` model for the manifest
- Add the manifest target to the list of project targets when loading its model

### Test Plan ⚒

- Verify unit tests pass via `swift test`
- Verify acceptance tests pass via `bundle exec rake features`
- Generate any fixture and validate the `Manifest` target is still generated
- Verify `Project.swift` is located within a top level "Manifest" group
- Verify autocomplete still work within the `Project.swift` file

### CODE REVIEW

File: Tests/TuistKitTests/Generator/ManifestTargetGeneratorTests.swift [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        // Then
+        XCTAssertEqual(target.name, "MyProject-Manifest")
+        XCTAssertEqual(target.sources.map { $0.asString }, ["/test/Project.swift"])
+        XCTAssertNil(target.infoPlist)
+        assertValidManifestBuildSettings(for: target)
+    }
+
+    // MARK: - Helpers
+
+    func assertValidManifestBuildSettings(for target: Target) {
```
Comment (marciniwanicki) [inline Tests/TuistKitTests/Generator/ManifestTargetGeneratorTests.swift]: I'd make it `private`, and maybe mark the class (and mocks) as `final` (to help the compiler move faster - at least in theory).
Maybe it's worth to change the locations a bit, currently everything is in `/test` and does not reflect the real world cases and it's harder to understand where does the `XCTAssertEqual(settings.base["FRAMEWORK_SEARCH_PATHS"], "/test")` path come from. Not sure, what about "/products/ProjectDescription.dylib" and maybe the path can be passed to the assertion as well?

File: Sources/TuistKit/Generator/ManifestTargetGenerator.swift [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    func generateManifestTarget(for project: String,
+                                at path: AbsolutePath) throws -> Target {
+        let settings = Settings(base: try manifestTargetBuildSettings(),
+                                debug: nil,
+                                release: nil)
+        let manifest = try manifestLoader.manifestPath(at: path, manifest: .project)
+        return Target(name: "\(project)-Manifest",
+                      platform: .macOS,
+                      product: .framework,
+                      bundleId: "",
```
Comment (ollieatkinson) [inline Sources/TuistKit/Generator/ManifestTargetGenerator.swift]: Do we need to set a bundle identifier ?

File: Sources/TuistKit/Models/Target.swift [hunk 3]
```diff
@@ -14,7 +14,7 @@ class Target: Equatable {
     let platform: Platform
     let product: Product
     let bundleId: String
-    let infoPlist: AbsolutePath
+    let infoPlist: AbsolutePath?
```
Comment (ollieatkinson) [inline Sources/TuistKit/Models/Target.swift]: Maybe it's worth a small in-line comment to explain why this is an optional

File: Sources/TuistKit/Generator/ProjectFileElements.swift [hunk 4]
```diff
@@ -123,7 +123,10 @@ class ProjectFileElements {
         }
 
         // Support files
-        files.insert(target.infoPlist)
+        if let infoPlist = target.infoPlist {
```
Comment (pepibumur) [inline Sources/TuistKit/Generator/ProjectFileElements.swift]: Is Xcode happy about having a framework without a `Info.plist`?

File: Sources/TuistKit/Models/Target.swift (hunk 3 above)
Comment (pepibumur) [reply Sources/TuistKit/Models/Target.swift]: Agree

File: Tests/TuistKitTests/Generator/ManifestTargetGeneratorTests.swift (hunk 1 above)
Comment (pepibumur) [reply Tests/TuistKitTests/Generator/ManifestTargetGeneratorTests.swift]: > I'd make it private, and maybe mark the class (and mocks) as final (to help the compiler move faster - at least in theory).

Agree here.

> Maybe it's worth to change the locations a bit, currently everything is in /test and does not reflect the real world cases and it's harder to understand where does the XCTAssertEqual(settings.base["FRAMEWORK_SEARCH_PATHS"], "/test") path come from.

I think it's ok in this case to keep the path `/test`. Although we know `/test` won't be the real path, we know for sure that `resourceLocator` will gives us that information, and that's why we have a mock for it.

File: Sources/TuistKit/Generator/ProjectFileElements.swift (hunk 4 above)
Comment (kwridan) [reply Sources/TuistKit/Generator/ProjectFileElements.swift]: Surprisingly yes, that seems to be the case for the Swift PM package description target:

tuist-packageDescription

File: Sources/TuistKit/Generator/ManifestTargetGenerator.swift (hunk 2 above)
Comment (kwridan) [reply Sources/TuistKit/Generator/ManifestTargetGenerator.swift]: I double checked what SwiftPM does for the package description target - it leaves it unset (i.e. `nil`) - might need to make this optional too at this level, I opted for blank to minimise disruptions to the API - perhaps an alternate could be `io.tuist.manifest.\(project)` or even `io.tuist.manifest.${PRODUCT_NAME:rfc1034identifier}`

**Comment by kwridan** [comment]
While testing I did come across this error:

```
error: unable to open dependencies file ( .../Build/Intermediates.noindex/MainApp.build/Debug/MainApp-Manifest.build/Objects-normal/x86_64/MainApp_Manifest_dependency_info.dat)Command Ld emitted errors but did not return a nonzero exit code to indicate failure
```

it doesn't fail the build, but does show up as an error, seems to be happening on master too - will investigate seeing I'm in the area.

**Comment by kwridan** [comment]
Looks like the same happens with the Swift PM package description target - will raise a separate issue to look further into this.

**Pull request merged**